File: 문제풀이/온라인 저지/백준/# 12. Gold Ⅳ/# 02110. 공유기 설치/python/main.py
import sys


def solution():
    n, c, adr = input()
    return parametric_search(adr, c, n)


def input():
    src = sys.stdin.buffer
    n, c = map(int, src.readline().split())
    adr = [int(x) for x in src.read().splitlines()]
    adr.sort()
    return n, c, adr


def parametric_search(adr, c, n):
    answer = 0
    lt = 1
    rt = adr[-1] - adr[0]
    while lt <= rt:
        cnt = 1
        mid = (lt + rt) >> 1
        before = adr[0]
        for i in range(1, n):
            if cnt == c:
                break
            if adr[i] - before >= mid:
                before = adr[i]
                cnt += 1
        if cnt == c:
            answer = mid
            lt = mid + 1
        else:
            rt = mid - 1
    return answer


print(solution())

File: 문제풀이/온라인 저지/백준/# 12. Gold Ⅳ/# 02110. 공유기 설치/kotlin/Main.kt
import java.util.*

fun main() {
    val n = i()
    val c = i()
    val arr = IntArray(n)
    for (i in 0 until n) {
        arr[i] = i()
    }
    Arrays.sort(arr)
    var lt = 1
    var rt = arr[n-1] - arr[0]
    var mid: Int
    var cnt: Int
    var answer = 1
    var success : Boolean
    var before: Int
    while (lt <= rt) {
        mid = (lt + rt) shr 1
        cnt = 0
        success = false
        before = arr[0] - mid
        for (e in arr) {
            if (e - before >= mid) {
                cnt ++
                before = e
            }
            if (cnt >= c) {
                success = true
                break
            }
        }
        if (success) {
            answer = mid
            lt = mid + 1
        } else {
            rt = mid - 1
        }
    }
    print(answer)
}

private const val S = 65536
private val iS = java.io.DataInputStream(System.`in`)
private val b = ByteArray(S)
private var c = 0
private var l = 0

private fun i(): Int {
    var v = 0
    var c = r()
    do v = v * 10 + c - 48 while (r().also { c = it } > 47)
    return v
}

private fun r(): Byte {
    if (c == l) {
        l = iS.read(b, 0.also { c = it }, S)
        if (l == -1) b[0] = -1
    }
    return b[c++]
}